File: main/binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public IList<IList<int>> ZigzagLevelOrder(TreeNode root) {
        var ret = new List<IList<int>>();
        if (root == null) return ret;

        var deque = new LinkedList<TreeNode>();
        deque.AddLast(root);

        while (deque.Count != 0) {
            var row = new List<int>(); // left to right
            for (var width = deque.Count; width != 0; --width) {
                root = deque.First.Value;
                deque.RemoveFirst();

                row.Add(root.val);
                if (root.left != null) deque.AddLast(root.left);
                if (root.right != null) deque.AddLast(root.right);
            }
            ret.Add(row);

            if (deque.Count == 0) break;

            row = new List<int>(); // right to left
            for (var width = deque.Count; width != 0; --width) {
                root = deque.Last.Value;
                deque.RemoveLast();

                row.Add(root.val);
                if (root.right != null) deque.AddFirst(root.right);
                if (root.left != null) deque.AddFirst(root.left);
            }
            ret.Add(row);
        }

        return ret;
    }
}

File: main/binary-tree-zigzag-level-order-traversal/binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        ret = []
        if root is None:
            return ret

        q = collections.deque()
        q.append(root)

        while q:
            row = []
            for _ in range(len(q)):
                root = q.popleft()
                row.append(root.val)
                if root.left is not None:
                    q.append(root.left)
                if root.right is not None:
                    q.append(root.right)
            ret.append(row)

            if not q:
                break

            row = []
            for _ in range(len(q)):
                root = q.pop()
                row.append(root.val)
                if root.right is not None:
                    q.appendleft(root.right)
                if root.left is not None:
                    q.appendleft(root.left)
            ret.append(row)

        return ret
